packages/react-hooks/src/useRefEffect.test.tsx

```diff
@@ -0,0 +1,62 @@
+import * as React from 'react';
+import { ReactWrapper, mount } from 'enzyme';
+import { useRefEffect } from './useRefEffect';
+
+describe('useRefEffect', () => {
+  // Use a jest.fn to log the calls to callback and cleanup
+  const log = jest.fn<void, [/*event:*/ 'callback' | 'cleanup', /*as:*/ 'div' | 'span', /*ele:*/ HTMLElement]>();
+  afterEach(() => log.mockClear());
+
+  let ref: React.RefObject<HTMLElement>;
+  const TestComponent: React.FunctionComponent<{ as: 'div' | 'span' }> = props => {
+    let setRef;
+    [ref, setRef] = useRefEffect<HTMLElement>(ele => {
+      log('callback', props.as, ele);
+      return () => log('cleanup', props.as, ele);
+    });
+    return <props.as ref={setRef} />;
+  };
+
+  let wrapper: ReactWrapper;
+
+  it('sets ref.current and calls callback, when the component is mounted', () => {
+    wrapper = mount(<TestComponent as="div" />);
+
+    expect(ref.current).toBeInstanceOf(HTMLDivElement);
+    expect(log).toHaveBeenCalledTimes(1);
+    expect(log).toHaveBeenCalledWith('callback', 'div', ref.current);
+  });
+
+  it('updates ref.current and calls cleanup and callback, when the ref changes', () => {
+    const prevRef = ref.current;
+
+    // Re-render as a <span /> to cause the ref to change
+    wrapper.setProps({ as: 'span' });
+
+    expect(ref.current).not.toBe(prevRef);
+    expect(ref.current).toBeInstanceOf(HTMLSpanElement);
+    expect(log).toHaveBeenCalledTimes(2);
+    expect(log).toHaveBeenNthCalledWith(1, 'cleanup', 'div', prevRef);
+    expect(log).toHaveBeenNthCalledWith(2, 'callback', 'span', ref.current);
+  });
+
+  it('does not call callback on render, if the ref has not changed', () => {
+    const prevRef = ref.current;
+
+    // Re-render without changing the type
+    wrapper.setProps({});
+
+    expect(ref.current).toBe(prevRef);
+    expect(log).not.toHaveBeenCalled();
+  });
+
+  it('clears ref.current and calls cleanup when the component is unmounted', () => {
+    const prevRef = ref.current;
+
+    wrapper.unmount();
```

Comment:
hmm we shouldn't write individual test this way which depends on the previous test to run. i always expect i can run one test and it should be pass by itself. in your case it's better to combine these into one test. adding comments between the steps will help readability